File: web/sc/js/display.js
/*! Single-Crystal Diffraction · © 2026 Miloš Dubajić · MIT · https://github.com/dubajicmilos/xrays-on-detector */

/**
 * How an intensity becomes something you can see.
 * A port of single_crystal/display.py, so the desktop app and this one show
 * the same picture for the same numbers.
 *
 * A diffraction pattern spans many decades. Plotted linearly, one reflection
 * is white and the rest are black; that is true to the data and useless to
 * look at. The stretch below maps intensity onto a fixed number of decades
 * below the strongest reflection, with a gain that slides the window.
 */

/** How many decades below the brightest reflection stay visible. */
export const DECADES = 6;

/**
 * Map intensity to 0..1 for display.
 *
 * `iMax` pins the normalisation to a chosen reference instead of the strongest
 * reflection present. Pass the strongest of the whole structure when stepping
 * through layers, or an upper layer is renormalised to its own weak maximum
 * and looks as bright as the zero layer.
 */
export function stretch(intensity, { gain = 1, log = true, iMax = null } = {}) {
  const n = intensity.length;
  const out = new Float64Array(n);
  let top = iMax;
  if (top === null) {
    top = 0;
    for (let i = 0; i < n; i++) if (intensity[i] > top) top = intensity[i];
  }
  if (!(top > 0)) return out;

  const floor = Math.pow(10, -DECADES);
  for (let i = 0; i < n; i++) {
    let v = (intensity[i] / top) * gain;
    v = v < 0 ? 0 : v > 1 ? 1 : v;
    if (log) {
      if (v < floor) v = floor;
      v = (Math.log10(v) + DECADES) / DECADES;
    }
    out[i] = v;
  }
  return out;
}

/** Marker radius in pixels from the stretched value. */
export const spotRadius = (value, scale = 9, floor = 1.1) =>
  floor + scale * value;

/**
 * [1, -1, 0] -> "11̅0", with an overbar on negative indices.
 *
 * U+0305 attaches to the character before it, so a two-digit index needs one
 * after every digit: -11 written as "11̅" bars only the second digit and reads
 * as the pair (1, -1). Barring both gives an unbroken line over "11".
 * Indices of ten and above are also spaced apart, or "1 12 0" reads as four.
 */
export function formatHkl(hkl) {
  const wide = hkl.some((v) => Math.abs(v) > 9);
  return hkl
    .map((v) => {
      const digits = String(Math.abs(v));
      return v < 0 ? [...digits].map((d) => d + "̅").join("") : digits;
    })
    .join(wide ? " " : "");
}

/** A round number near span/6, for a scale bar or a grid. */
export function niceStep(span) {
  if (!(span > 0)) return 1;
  const raw = span / 6;
  const mag = Math.pow(10, Math.floor(Math.log10(raw)));
  for (const m of [1, 2, 2.5, 5, 10]) if (raw <= m * mag) return m * mag;
  return 10 * mag;
}

/** Sample a flat 256*3 colormap, returning "rgb(r,g,b)". */
export function sampleLut(lut, value) {
  let i = Math.round(value * 255);
  i = i < 0 ? 0 : i > 255 ? 255 : i;
  return `rgb(${lut[3 * i]},${lut[3 * i + 1]},${lut[3 * i + 2]})`;
}

File: single_crystal/display.py
"""How an intensity becomes something you can see.

Shared by the desktop app and, in ported form, by web/sc/js/render.js, so the
two show the same picture for the same numbers.

A diffraction pattern spans many decades. Plotted linearly, one reflection is
white and the rest are black; that is true to the data and useless to look at.
The stretch below maps intensity onto a fixed number of decades below the
strongest reflection, with a gain that slides the window up and down. It is
the same treatment the Game of Diffraction applies to its detector image.
"""
from __future__ import annotations

import math

import numpy as np

#: How many decades below the brightest reflection stay visible.
DECADES = 6.0


def stretch(intensity, gain: float = 1.0, log: bool = True, i_max: float | None = None):
    """Map intensity to 0..1 for display.

    gain multiplies the normalised intensity before the stretch, so raising it
    brings weak reflections up; the brightest simply saturates.

    i_max pins the normalisation to a chosen reference instead of the strongest
    reflection present. Pass the strongest reflection of the whole structure
    when stepping through layers, or an upper layer will be renormalised to its
    own weak maximum and look as bright as the zero layer.
    """
    I = np.asarray(intensity, dtype=float)
    if I.size == 0:
        return I
    top = float(I.max()) if i_max is None else float(i_max)
    if not (top > 0):
        return np.zeros_like(I)

    v = np.clip(I / top * gain, 0.0, 1.0)
    if not log:
        return v
    floor = 10.0 ** (-DECADES)
    v = np.clip(v, floor, 1.0)
    return (np.log10(v) + DECADES) / DECADES


def spot_sizes(display, scale: float = 180.0, floor: float = 3.0):
    """Marker area for a scatter plot, from the stretched value.

    Area rather than radius tracks the eye's reading of a spot, and a floor
    keeps a just-visible reflection from vanishing into a subpixel dot.
    """
    d = np.asarray(display, dtype=float)
    return floor + scale * d


def format_hkl(hkl, bar: bool = True) -> str:
    """(1, -1, 0) -> '11̅0'. Negative indices take an overbar, as they are
    written everywhere else.

    U+0305 attaches to the character before it, so a two-digit index needs one
    after every digit: -11 written as '11̅' bars only the second digit and reads
    as the pair (1, -1). Barring both gives an unbroken line over '11'.
    """
    out = []
    for v in hkl:
        v = int(v)
        digits = str(abs(v))
        if v < 0 and bar:
            out.append("".join(d + "̅" for d in digits))
        else:
            out.append(str(v))
    # Indices of ten and above need a separator or 1 12 0 reads as 1 1 2 0.
    sep = " " if any(abs(int(v)) > 9 for v in hkl) else ""
    return sep.join(out)


def nice_step(span: float) -> float:
    """A round number near span/6, for a scale bar or a grid."""
    if span <= 0:
        return 1.0
    raw = span / 6.0
    mag = 10.0 ** math.floor(math.log10(raw))
    for m in (1.0, 2.0, 2.5, 5.0, 10.0):
        if raw <= m * mag:
            return m * mag
    return 10.0 * mag
